pkg/operation/controlplanesecrets/secrets_generator.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplanesecrets
+
+import (
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	"github.com/gardener/gardener/pkg/utils/infodata"
+	"github.com/gardener/gardener/pkg/utils/secrets"
+)
+
+// SecretsGenerator extends the SecretsManager with methods which can generate InfoData from the given secret configurations.
+type SecretsGenerator struct {
+	*SecretsManager
+}
+
+// NewSecretsGenerator creates a new SecretsGenerator
+func NewSecretsGenerator(
+	gardenerResourceDataList gardencorev1alpha1helper.GardenerResourceDataList,
+	apiserverBasicAuthConfig secrets.ConfigInterface,
+	staticTokenConfig secrets.ConfigInterface,
+	certificateAuthorityConfigs map[string]*secrets.CertificateSecretConfig,
+	secretConfigGenerator SecretConfigGeneratorFunc,
+) *SecretsGenerator {
+	s := &SecretsGenerator{&SecretsManager{}}
+	s.CertificateAuthorities = map[string]*secrets.Certificate{}
+	s.CertificateAuthorities = map[string]*secrets.Certificate{}
+	s.GardenerResourceDataList = gardenerResourceDataList
+	s.APIServerBasicAuthConfig = apiserverBasicAuthConfig
+	s.StaticTokenConfig = staticTokenConfig
+	s.CertificateAuthorityConfigs = certificateAuthorityConfigs
+	s.SecretConfigGenerator = secretConfigGenerator
+	return s
+}
+
+// GenerateAPIServerBasicAuth will generate InfoData for the APIServerBasicAuth
+func (s *SecretsGenerator) GenerateAPIServerBasicAuth() error {
```

Comment:
I think we could have a single exported method called `Generate` which is doing the work of the methods bellow. We just return the errors in the client code, anyways. It then will have another boolean field to determine if it should generate basic auth secrets, called `withBasicAuth` for example.

Reply:
I had this idea as well. Generally could look something like (same for generator, deployer and loader):
```
secretsManager := NewSecretsManager().WithStaticToken(...).WithGeneratorFunc(...).WithCAConfigs(...)
if shootWantsBasicAuth {
    secretsManager := secretsManager.WithAPIServerBasicAuth(...)
}
secretsDeployer := NewSecretsDeployer().WithManager(secretsManager).WithExistingSecrets(...)
if err := secretsDeployer.Deploy(); err != nil {
...
}
mutex.Lock()
defer mutex.Unlock()
for name, secret := range secretsDeployer {
  b.Secrets[name] = secret
}
return nil
```

Might as well change the `SecretsManager` to `SecretsConfig` I think something like that makes more sense.

Reply:
Agree with @swilen-iwanow, and I would say perhaps the same is true for `Deploy`, and `Load`. See my other proposal about combining everything together in one type, e.g. `SecretsManager`. This type could then expose methods such as `Generate`, `Deploy`, and `Load`, and many of the current methods would no longer be exported.